### PR TITLE
Replace the SearchContext with QueryShardContext when building aggregator factories

### PR DESCRIPTION
**Pull request by jimczi**
This commit replaces the `SearchContext` with the `QueryShardContext` when building aggregator factories.
Aggregator factories are part of the `SearchContext` so they shouldn't require a `SearchContext` to create them.The main changes here are the signatures of `AggregationBuilder#build` that now takes a `QueryShardContext` and `AggregatorFactory#createInternal` that passes the `SearchContext` to build the `Aggregator`.

Relates #46523

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java [hunk 1]
```diff
@@ -551,9 +551,11 @@ private static Response prepareRamIndex(Request request,
                     ParsedDocument parsedDocument = indexService.mapperService().documentMapper().parse(sourceToParse);
                     indexWriter.addDocuments(parsedDocument.docs());
                     try (IndexReader indexReader = DirectoryReader.open(indexWriter)) {
+                        final IndexSearcher searcher = new IndexSearcher(indexReader);
+                        searcher.setQueryCache(null);
                         final long absoluteStartMillis = System.currentTimeMillis();
                         QueryShardContext context =
-                            indexService.newQueryShardContext(0, indexReader, () -> absoluteStartMillis, null);
+                            indexService.newQueryShardContext(0, searcher, () -> absoluteStartMillis, null);
```
Comment (javanna) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java]: out of curiosity: was this change necessary?

**Review by javanna — APPROVED** [review]
the idea makes sense to me, it is not trivial to follow all that happens in this PR due to its size. I trust you though :)

**Comment by jimczi** [comment]
run elasticsearch-ci/1

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java (hunk 1 above)
Comment (jimczi) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java]: For simplicity the `QueryShardContext` should expose an `IndexSearcher` that is configured with the correct similarity, cache, ... I can make this change in a different pr if you like ?

Comment (javanna) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java]: it would be nice, unless that cause too much extra work on your end.

Comment (jimczi) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java]: I opened https://github.com/elastic/elasticsearch/pull/46584 to split this pr a bit.

Comment (jimczi) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java]: I merged #46584 and incorporated the changes in this pr, can you take another look ?

**Review by javanna — APPROVED** [review]
still LGTM thanks @jimczi
